### PR TITLE
Refactor mirror command

### PR DESCRIPTION
**Pull request by lucklove**
### What problem does this PR solve? 
https://github.com/pingcap/tiup/issues/814

### What is changed and how it works?
- [x] Support publish to local mirror
- [x] Support merge mirror
- [x] Refactor tiup-server
- [x] Support add owner
- [x] Add test

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release notes:

```release-note
Support merge offline packages
```

### CODE REVIEW

File: cmd/mirror.go [hunk 1]
```diff
@@ -221,80 +118,192 @@ func newMirrorSetCmd() *cobra.Command {
 	return cmd
 }
 
+// the `mirror introduce` sub command
+func newMirrorIntroduceCmd() *cobra.Command {
+	name := ""
+	privPath := ""
+
+	cmd := &cobra.Command{
+		Use:   "introduce <id>",
```
Comment (lonng) [inline cmd/mirror.go:127]: How about using` tiup mirror grant/revoke` to add a new owner and remove exists owner.

File: cmd/mirror.go [hunk 2]
```diff
@@ -221,80 +118,192 @@ func newMirrorSetCmd() *cobra.Command {
...
+		Short: "introduce a new owner",
+		Long:  "introduce a new owner to current mirror",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return cmd.Help()
+			}
+
+			id := args[0]
+			if name == "" {
+				name = id
```
Comment (lonng) [inline cmd/mirror.go:137]: We assigned a default value for name, do we need to print some message to hint the user? e.g:
```go
fmt.Printf("The --name hasn't specified value, default to %s\n", id).
```

File: cmd/mirror.go [hunk 3]
```diff
@@ -221,80 +118,192 @@ func newMirrorSetCmd() *cobra.Command {
...
+				return err
+			}
+
+			env := environment.GlobalEnv()
+			return env.V1Repository().Mirror().Introduce(id, name, pubKey)
+		},
+	}
+
+	cmd.Flags().StringVarP(&name, "name", "n", "", "the name of the owner, default: id of the owner")
+	cmd.Flags().StringVarP(&privPath, "key", "k", "", "private/public key path of the owner")
```
Comment (lonng) [inline cmd/mirror.go:156]: I think we should make the description more clear if only `private key` can be assigned.
`private/public key path of the owner` -> `Specify the private key path of the owner`.

IMO, all descriptions for flags should be started with a upper case.

File: cmd/mirror.go [hunk 4]
```diff
@@ -221,80 +118,192 @@ func newMirrorSetCmd() *cobra.Command {
...
 	desc := ""
 	standalone := false
 	hidden := false
 	yanked := false
 
 	cmd := &cobra.Command{
-		Use:  "modify <component>[:version] [flags]",
-		Long: "modify component attributes (hidden, standalone, yanked)",
+		Use:   "modify <component>[:version] [flags]",
+		Short: "modify published component",
```
Comment (lonng) [inline cmd/mirror.go:171]:
```suggestion
		Short: "Modify published component",
```

File: cmd/mirror.go [hunk 5]
```diff
@@ -221,80 +118,192 @@ func newMirrorSetCmd() *cobra.Command {
...
+		Long:  "modify component attributes (hidden, standalone, yanked)",
```
Comment (lonng) [inline cmd/mirror.go:172]:
```suggestion
		Long:  "Modify component attributes (hidden, standalone, yanked)",
```

File: cmd/mirror.go [hunk 6]
```diff
@@ -27,18 +28,16 @@ import (
 	"github.com/pingcap/tiup/pkg/environment"
 	"github.com/pingcap/tiup/pkg/localdata"
 	"github.com/pingcap/tiup/pkg/repository"
-	"github.com/pingcap/tiup/pkg/repository/remote"
+	"github.com/pingcap/tiup/pkg/repository/model"
+	ru "github.com/pingcap/tiup/pkg/repository/utils"
+	"github.com/pingcap/tiup/pkg/repository/v0manifest"
```
Comment (lonng) [inline cmd/mirror.go:33]: I think we can remove the `v0manifest` in the `v1.4.0`. @AstroProfundis

File: cmd/mirror.go [hunk 7]
```diff
@@ -310,72 +319,60 @@ func newMirrorPublishCmd() *cobra.Command {
...
-			if privPath == "" {
-				privPath = env.Profile().Path(localdata.KeyInfoParentDir, "private.json")
+			hashes, length, err := ru.HashFile(tarpath)
+			if err != nil {
+				return err
 			}
 
-			// Get the private key
-			f, err := os.Open(privPath)
+			tarfile, err := os.Open(args[2])
```
Comment (lonng) [inline cmd/mirror.go:333]: Please use the variable declared at L322.

File: cmd/mirror.go [hunk 8]
```diff
@@ -310,72 +319,60 @@ func newMirrorPublishCmd() *cobra.Command {
...
 			if err != nil {
-				return err
+				return errors.Annotate(err, "open tarbal")
```
Comment (lonng) [inline cmd/mirror.go:335]:
```suggestion
				return errors.Annotatef(err, "open tarball: %s", tarpath)
```

File: cmd/mirror.go (hunk 1 above)
Comment (lucklove) [reply cmd/mirror.go:127]: I don't think remove an owner is possible. The owner doesn't has a 'yank' field so we can just remove it from the manifest file. After that, the existed components owned by that owner can't be verified (since the key is removed with that owner) and there will be signature error when user execute `tiup list` or execute the installed component.

File: cmd/mirror.go (hunk 2 above)
Comment (lucklove) [reply cmd/mirror.go:137]: OK

File: cmd/mirror.go (hunk 1 above)
Comment (lucklove) [reply cmd/mirror.go:127]: Maybe we should ignore the components missing owner, and report warning when user already install that component.

Comment (lucklove) [reply cmd/mirror.go:127]: But what if something depends on it? 

IMO, delete owner directly breaks the `yank` mechanism

File: cmd/mirror.go (hunk 6 above)
Comment (AstroProfundis) [reply cmd/mirror.go:33]: I agree, https://github.com/pingcap/tiup/issues/906

File: cmd/mirror.go [hunk 9]
```diff
@@ -221,80 +118,193 @@ func newMirrorSetCmd() *cobra.Command {
...
+		Short: "introduce a new owner",
+		Long:  "introduce a new owner to current mirror",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) < 1 {
+				return cmd.Help()
+			}
+
+			id := args[0]
+			if name == "" {
+				fmt.Printf("The --name hasn't specified value, default to %s\n", id)
```
Comment (AstroProfundis) [inline cmd/mirror.go:137]:
```suggestion
				fmt.Printf("The --name is not specified, using %s as default\n", id)
```

File: cmd/mirror.go [hunk 10]
```diff
@@ -221,80 +118,193 @@ func newMirrorSetCmd() *cobra.Command {
...
+			comp, ver := environment.ParseCompVersion(component)
 			m, err := env.V1Repository().FetchComponentManifest(comp, true)
 			if err != nil {
 				return err
 			}
 
-			if endpoint == "" {
-				endpoint = environment.Mirror()
+			v1manifest.RenewManifest(m, time.Now())
+			m.Version++
```
Comment (AstroProfundis) [inline cmd/mirror.go:190]: Why not bump version in `RenewManifest()`?

File: pkg/cluster/audit/audit.go [hunk 11]
```diff
@@ -81,7 +82,7 @@ func ShowAuditList(dir string) error {
 
 // OutputAuditLog outputs audit log.
 func OutputAuditLog(dir string, data []byte) error {
-	fname := filepath.Join(dir, base52.Encode(time.Now().UnixNano()))
+	fname := filepath.Join(dir, base52.Encode(time.Now().UnixNano()+rand.Int63n(1000)))
```
Comment (AstroProfundis) [inline pkg/cluster/audit/audit.go:85]: Better split this to another PR

File: pkg/repository/merge_mirror.go [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+	name          string
+	componentItem v1manifest.ComponentItem
+	versionItem   v1manifest.VersionItem
+	version       string
+	os            string
+	arch          string
+	desc          string
+}
+
+// returns component that exists in source but not in base
```
Comment (AstroProfundis) [inline pkg/repository/merge_mirror.go:38]: "source" and "base" are bad names, they have similar meaning and might be confusing. how about "base" and "input" or "addition"

File: pkg/repository/merge_mirror.go [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+
+		baseComponent, err := fetchComponentManifestFromMirror(base, name)
+		if err != nil {
+			return nil, err
+		}
+		sourceComponent, err := fetchComponentManifestFromMirror(source, name)
+		if err != nil {
+			return nil, err
+		}
+		if sourceComponent == nil {
```
Comment (AstroProfundis) [inline pkg/repository/merge_mirror.go:67]: Is this possible?

File: pkg/repository/merge_mirror.go [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+		for ver, verinfo := range versions {
+			// Don't merge nightly this time
+			if v0manifest.Version(ver).IsNightly() {
+				continue
+			}
+
+			// this version not exits in base
+			if baseManifest.VersionListWithYanked(plat)[ver].URL == "" {
+				osArch := strings.Split(plat, "/")
+				if len(osArch) != 2 {
```
Comment (AstroProfundis) [inline pkg/repository/merge_mirror.go:92]: What's this check for?

File: pkg/repository/merge_mirror.go [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+	defer r.Close()
+
+	snap := v1manifest.Snapshot{}
+	if _, err := v1manifest.ReadNoVerify(r, &snap); err != nil {
+		return nil, err
+	}
+
+	v := snap.Meta[fmt.Sprintf("/%s.json", component)].Version
+	if v == 0 {
+		return nil, nil
```
Comment (AstroProfundis) [inline pkg/repository/merge_mirror.go:181]: I think this is not reasonable to return `nil, nil`

File: pkg/repository/merge_mirror.go [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+	name, ver, entry, os, arch, desc string,
+	filehash v1manifest.FileHash) *v1manifest.Component {
+	initTime := time.Now()
+
+	// update manifest
+	if m == nil {
+		m = v1manifest.NewComponent(name, desc, initTime)
+	} else {
+		v1manifest.RenewManifest(m, initTime)
+		m.Version++
```
Comment (AstroProfundis) [inline pkg/repository/merge_mirror.go:263]: ditto, bumping version is not atomic here

**Review by AstroProfundis — COMMENTED** [review]
And don't change the file mode of `pkg/cluster/executor/executor.go`

File: pkg/repository/mirror.go [hunk 17]
```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+
+	if info.Filename() != "" {
+		tarAddr := fmt.Sprintf("%s/api/v1/tarball/%s", l.Source(), sid)
+		resp, err := utils.PostFile(info, tarAddr, "file", info.Filename())
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode >= 300 {
```
Comment (AstroProfundis) [inline pkg/repository/mirror.go:346]: What about 301, 302, 303 and 307?

File: pkg/repository/mirror.go [hunk 18]
```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+	}
+
+	payload, err := json.Marshal(manifest)
+	if err != nil {
+		return err
+	}
+	bodyBuf := bytes.NewBuffer(payload)
+	qpairs := []string{}
+	if info.Yanked() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "yanked", *info.Yanked()))
```
Comment (AstroProfundis) [inline pkg/repository/mirror.go:358]: Maybe it's better to define those keys as global vars or consts

File: pkg/repository/mirror.go [hunk 19]
```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+	}
+	if info.Standalone() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "standalone", *info.Standalone()))
+	}
+	if info.Hidden() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "hidden", *info.Hidden()))
+	}
+	qstr := ""
+	if len(qpairs) > 0 {
+		qstr = "?" + strings.Join(qpairs, "&")
```
Comment (AstroProfundis) [inline pkg/repository/mirror.go:368]: Better use `net/url.Values.Encode()`

File: pkg/repository/mirror.go [hunk 20]
```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+	}
+	manifestAddr := fmt.Sprintf("%s/api/v1/component/%s/%s%s", l.Source(), sid, manifest.Signed.(*v1manifest.Component).ID, qstr)
+
+	resp, err := http.Post(manifestAddr, "text/json", bodyBuf)
```
Comment (AstroProfundis) [inline pkg/repository/mirror.go:372]: The default HTTP client has no timeout set, we should create a client with proper timeout setting and use it for requests.

File: pkg/repository/mirror.go [hunk 21]
```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
...
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 300 {
```
Comment (AstroProfundis) [inline pkg/repository/mirror.go:378]: ditto, what about other redirect codes?

File: pkg/repository/model/model.go [hunk 22]
```diff
@@ -0,0 +1,406 @@
...
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/juju/errors"
+	"github.com/pingcap/tiup/pkg/logger/log"
+	"github.com/pingcap/tiup/pkg/repository/store"
+	"github.com/pingcap/tiup/pkg/repository/v1manifest"
+	"github.com/pingcap/tiup/pkg/utils"
+)
+
+// Model defines operations on the manifests
+type Model interface {
```
Comment (AstroProfundis) [inline pkg/repository/model/model.go:29]: I think it could be something like `model.Filesystem` or `model.Storage` instead of `model.Model`

File: pkg/repository/store/txn.go [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+	filepath := t.store.path(filename)
+	if utils.IsExist(path.Join(t.root, filename)) {
+		filepath = path.Join(t.root, filename)
+	}
+	var wc io.ReadCloser
+	if file, err := os.Open(filepath); err == nil {
+		wc = file
+	} else if os.IsNotExist(err) && t.store.upstream != "" {
+		url := fmt.Sprintf("%s/%s", t.store.upstream, filename)
+		if resp, err := http.Get(url); err == nil {
```
Comment (AstroProfundis) [inline pkg/repository/store/txn.go:137]: Set timeout

**Review by AstroProfundis — COMMENTED** [review]
We should add more comments in `txn.go` to describe how the merging process work.

File: cmd/mirror.go (hunk 10 above)
Comment (lucklove) [reply cmd/mirror.go:190]: good idea

File: pkg/cluster/audit/audit.go (hunk 11 above)
Comment (lucklove) [reply pkg/cluster/audit/audit.go:85]: https://github.com/pingcap/tiup/pull/916

File: pkg/repository/mirror.go (hunk 17 above)
Comment (lucklove) [reply pkg/repository/mirror.go:346]: It's unexpected, the server shouldn't return this.

**Review by AstroProfundis — APPROVED** [review]
